pkg/ui/workspaces/cluster-ui/src/insights/utils.ts

```diff
@@ -420,3 +422,75 @@ export function getTxnInsightRecommendations(
   );
   return recs;
 }
+
+// The functions below de-duplicate insights of the same type for a given fingerprint.
+// The front-end requires de-duplicated insights for the fingerprints pages because the backend has a problem+cause
+// definition for insight types, and the SQL API query selects distinct on fingerprint, problem, and cause.
+
+export function getStmtFingerprintInsightCounts(
+  insightEvents: StmtInsightEvent[],
+): ExecutionInsightCountEvent[] {
+  if (!insightEvents) {
+    return null;
+  }
+  const result: ExecutionInsightCountEvent[] = [];
+  const stmtInsightMap = new Map<string, Set<InsightNameEnum>>();
+  insightEvents.forEach(event => {
+    if (!stmtInsightMap.has(event.statementFingerprintID)) {
+      const stmtInsightTypes = new Set<InsightNameEnum>();
+      event.insights.forEach(insight => stmtInsightTypes.add(insight.name));
+      stmtInsightMap.set(event.statementFingerprintID, stmtInsightTypes);
+      result.push({
+        fingerprintID: event.statementFingerprintID,
+        insightCount: event.insights.length,
+      });
+    } else {
+      const mapValues = stmtInsightMap.get(event.statementFingerprintID);
+      event.insights.forEach(insight => {
+        if (!mapValues.has(insight.name)) {
+          mapValues.add(insight.name);
+          const idx = result.findIndex(
+            item => item.fingerprintID === event.statementFingerprintID,
+          );
+          result[idx].insightCount += 1;
+        }
+      });
+    }
```

Comment:
I might be reading this wrong but is this mapping fingerprints -> insight count by mapping fingerprint -> set of insight types?
It seems like it might not be doing what you're intending to.
Instead of having 2 branches, why not do smth like
```
const insightsSet = map[fingerprint] ? new set;
insightsSet.add(type);
map[fingerprint] = insightsSet;
```
Then, once you've gone over all the events, you can. map the entries in the map to {fingerprintID: fingerprint, count: sizeof(set))}